akka-actor/src/main/scala/akka/dispatch/Mailbox.scala

```diff
@@ -399,11 +399,9 @@ class NodeMessageQueue extends AbstractNodeQueue[Envelope] with MessageQueue wit
 
   final def hasMessages: Boolean = !isEmpty()
 
-  @tailrec final def cleanUp(owner: ActorRef, deadLetters: MessageQueue): Unit = {
-    val envelope = dequeue()
-    if (envelope ne null) {
-      deadLetters.enqueue(owner, envelope)
-      cleanUp(owner, deadLetters)
+  final def cleanUp(owner: ActorRef, deadLetters: MessageQueue): Unit = {
+    while (hasMessages) {
+      deadLetters.enqueue(owner, dequeue())
```

Comment:
The dequeue and look if that was the last message is no longer atomic which it was in the original implementation, that means that hasMessages could return true, but when deque is called it no longer is and we try to enqueue a null. In addition to that it now touches the queue twice which could have performance implications.

Reply:
@johanandren make sense. How about turning the recursive approach into

```scala
var envelope = dequeue()
while (envelope ne null) {
  deadLetters.enqueue(owner, envelope)
  envelope = dequeue()
}
```

Reply:
What would be the advantage of that? Recursion vs while might be a subjective style choice, but we often use recursion is often used like this in the Akka code base.

Reply:
I prefer the recursion as well. 

On a side note, if we believe the code needs to be simplified it's probably because its intention is not clear. If I understand @johanandren comment, the reasons it is as such are: atomicity and optimization (not touching the queue twice). 

Perhaps instead of changing the impl, we should add some comments explaining why this implementation choice. 

Now that @johanandren explain it, it becomes obvious to me. Without it I would have to read, think, think again and again and then come to that conclusion. Or missed it completely.

Reply:
Hard to draw the line, if we do there is an abundance of places you need to explain choices made for efficient concurrency. I am not convinced that is what we should prefer spending our time on.

Reply:
I'm not saying that we should scan the code base for similar things, but since we are talking about this here and we have an open PR we could improve it.

Reply:
Sure, I'm ok with that if contributor wants to add that

Reply:
I agree with not changing something like that for exclusively cosmetic reasons.

> The dequeue and look if that was the last message is no longer atomic which it was in the original implementation, that means that hasMessages could return true, but when deque is called it no longer is and we try to enqueue a null. In addition to that it now touches the queue twice which could have performance implications.

That's probably an accurate description of the current implementation but we actually don't know about the intention and whether it made or still makes sense. If there was a deeper intention behind it, it should have been documented. Most likely, though, it just doesn't matter (There's only a single reader to a mailbox, but there's probably a race condition that would allow new messages to be pushed while it is emptied here. Adding another `hasMessages` might actually miss fewer of those message but the race condition would still be there.)

> Hard to draw the line, if we do there is an abundance of places you need to explain choices made for efficient concurrency. I am not convinced that is what we should prefer spending our time on.

We should definitely document our intentions and assumptions more when it comes to writing efficient *and correct* concurrent code. For existing code, guessing the intentions and documenting those as facts seems much less useful, though :)